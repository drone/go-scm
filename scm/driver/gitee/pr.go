// Copyright 2017 Drone.IO Inc. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package gitee

import (
	"context"
	"fmt"
	"time"

	"github.com/drone/go-scm/scm"
)

type pullService struct {
	client *wrapper
}

func (s *pullService) Find(ctx context.Context, repo string, number int) (*scm.PullRequest, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d", repo, number)
	out := new(pr)
	res, err := s.client.do(ctx, "GET", path, nil, out)
	return convertPullRequest(out), res, err
}

func (s *pullService) FindComment(ctx context.Context, repo string, _ int, id int) (*scm.Comment, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/comments/%d", repo, id)
	out := new(prComment)
	res, err := s.client.do(ctx, "GET", path, nil, out)
	return convertPullRequestComment(out), res, err
}

func (s *pullService) List(ctx context.Context, repo string, opts scm.PullRequestListOptions) ([]*scm.PullRequest, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls?%s", repo, encodePullRequestListOptions(opts))
	out := []*pr{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertPullRequestList(out), res, err
}

func (s *pullService) ListChanges(ctx context.Context, repo string, number int, _ scm.ListOptions) ([]*scm.Change, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d/files", repo, number)
	out := []*prFile{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertPrChangeList(out), res, err
}

func (s *pullService) ListComments(ctx context.Context, repo string, number int, opts scm.ListOptions) ([]*scm.Comment, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d/comments/?%s", repo, number, encodeListOptions(opts))
	out := []*prComment{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertPullRequestComments(out), res, err
}

func (s *pullService) ListCommits(ctx context.Context, repo string, number int, _ scm.ListOptions) ([]*scm.Commit, *scm.Response, error) {
	path := fmt.Sprintf("/repos/%s/pulls/%d/commits", repo, number)
	out := []*prCommit{}
	res, err := s.client.do(ctx, "GET", path, nil, &out)
	return convertPrCommitList(out), res, err
}

func (s *pullService) Merge(ctx context.Context, repo string, number int) (*scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d/merge", repo, number)
	res, err := s.client.do(ctx, "PUT", path, nil, nil)
	return res, err
}

func (s *pullService) Close(ctx context.Context, repo string, number int) (*scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d", repo, number)
	data := map[string]string{"state": "closed"}
	res, err := s.client.do(ctx, "PATCH", path, &data, nil)
	return res, err
}

func (s *pullService) Create(ctx context.Context, repo string, input *scm.PullRequestInput) (*scm.PullRequest, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls", repo)
	in := &prInput{
		Title: input.Title,
		Body:  input.Body,
		Head:  input.Source,
		Base:  input.Target,
	}
	out := new(pr)
	res, err := s.client.do(ctx, "POST", path, in, out)
	return convertPullRequest(out), res, err
}

func (s *pullService) CreateComment(ctx context.Context, repo string, number int, input *scm.CommentInput) (*scm.Comment, *scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/%d/comments", repo, number)
	in := &prCommentInput{
		Body: input.Body,
	}
	out := new(prComment)
	res, err := s.client.do(ctx, "POST", path, in, out)
	return convertPullRequestComment(out), res, err
}

func (s *pullService) DeleteComment(ctx context.Context, repo string, _ int, id int) (*scm.Response, error) {
	path := fmt.Sprintf("repos/%s/pulls/comments/%d", repo, id)
	res, err := s.client.do(ctx, "DELETE", path, nil, nil)
	return res, err
}

type (
	pr struct {
		ID                int        `json:"id"`
		URL               string     `json:"url"`
		HtmlURL           string     `json:"html_url"`
		DiffURL           string     `json:"diff_url"`
		PatchURL          string     `json:"patch_url"`
		IssueURL          string     `json:"issue_url"`
		CommitsURL        string     `json:"commits_url"`
		ReviewCommentsURL string     `json:"review_comments_url"`
		ReviewCommentURL  string     `json:"review_comment_url"`
		CommentsURL       string     `json:"comments_url"`
		Number            int        `json:"number"`
		State             string     `json:"state"`
		Title             string     `json:"title"`
		Body              string     `json:"body"`
		AssigneesNumber   int        `json:"assignees_number"`
		TestersNumber     int        `json:"testers_number"`
		Assignees         []assignee `json:"assignees"`
		Testers           []tester   `json:"testers"`
		Milestone         milestone  `json:"milestone"`
		Labels            []label    `json:"labels"`
		Locked            bool       `json:"locked"`
		CreatedAt         time.Time  `json:"created_at"`
		UpdatedAt         time.Time  `json:"updated_at"`
		ClosedAt          time.Time  `json:"closed_at"`
		MergedAt          time.Time  `json:"merged_at"`
		Mergeable         bool       `json:"mergeable"`
		CanMergeCheck     bool       `json:"can_merge_check"`
		Head              headOrBase `json:"head"`
		Base              headOrBase `json:"base"`
		User              user       `json:"user"`
	}
	assignee struct {
		ID                int    `json:"id"`
		Login             string `json:"login"`
		Name              string `json:"name"`
		AvatarURL         string `json:"avatar_url"`
		URL               string `json:"url"`
		HtmlURL           string `json:"html_url"`
		Remark            string `json:"remark"`
		FollowersURL      string `json:"followers_url"`
		FollowingURL      string `json:"following_url"`
		GistsURL          string `json:"gists_url"`
		StarredURL        string `json:"starred_url"`
		SubscriptionsURL  string `json:"subscriptions_url"`
		OrganizationsURL  string `json:"organizations_url"`
		ReposURL          string `json:"repos_url"`
		EventsURL         string `json:"events_url"`
		ReceivedEventsURL string `json:"received_events_url"`
		Type              string `json:"type"`
	}
	tester struct {
		ID                int    `json:"id"`
		Login             string `json:"login"`
		Name              string `json:"name"`
		AvatarURL         string `json:"avatar_url"`
		URL               string `json:"url"`
		HtmlURL           string `json:"html_url"`
		Remark            string `json:"remark"`
		FollowersURL      string `json:"followers_url"`
		FollowingURL      string `json:"following_url"`
		GistsURL          string `json:"gists_url"`
		StarredURL        string `json:"starred_url"`
		SubscriptionsURL  string `json:"subscriptions_url"`
		OrganizationsURL  string `json:"organizations_url"`
		ReposURL          string `json:"repos_url"`
		EventsURL         string `json:"events_url"`
		ReceivedEventsURL string `json:"received_events_url"`
		Type              string `json:"type"`
	}
	milestone struct {
		URL          string      `json:"url"`
		HtmlURL      string      `json:"html_url"`
		ID           int         `json:"id"`
		Number       int         `json:"number"`
		RepositoryID interface{} `json:"repository_id"`
		State        string      `json:"state"`
		Title        string      `json:"title"`
		Description  string      `json:"description"`
		UpdatedAt    time.Time   `json:"updated_at"`
		CreatedAt    time.Time   `json:"created_at"`
		OpenIssues   int         `json:"open_issues"`
		ClosedIssues int         `json:"closed_issues"`
		DueOn        string      `json:"due_on"`
	}
	label struct {
		ID           int       `json:"id"`
		Name         string    `json:"name"`
		Color        string    `json:"color"`
		RepositoryID int       `json:"repository_id"`
		URL          string    `json:"url"`
		CreatedAt    time.Time `json:"created_at"`
		UpdatedAt    time.Time `json:"updated_at"`
	}
	prRepo struct {
		ID        int    `json:"id"`
		FullName  string `json:"full_name"`
		HumanName string `json:"human_name"`
		URL       string `json:"url"`
		//Namespace   namespace `json:"namespace"`
		Path        string   `json:"path"`
		Name        string   `json:"name"`
		Owner       user     `json:"owner"`
		Assigner    assignee `json:"assigner"`
		Description string   `json:"description"`
		Private     bool     `json:"private"`
		Public      bool     `json:"public"`
		Internal    bool     `json:"internal"`
		Fork        bool     `json:"fork"`
		HtmlURL     string   `json:"html_url"`
		SshURL      string   `json:"ssh_url"`
	}
	headOrBase struct {
		Label string `json:"label"`
		Ref   string `json:"ref"`
		Sha   string `json:"sha"`
		User  user   `json:"user"`
		Repo  prRepo `json:"repo"`
	}

	prComment struct {
		URL            string    `json:"url"`
		ID             int       `json:"id"`
		User           user      `json:"user"`
		CreatedAt      time.Time `json:"created_at"`
		UpdatedAt      time.Time `json:"updated_at"`
		Body           string    `json:"body"`
		HtmlURL        string    `json:"html_url"`
		PullRequestURL string    `json:"pull_request_url"`
	}

	prFile struct {
		Sha       string `json:"sha"`
		Filename  string `json:"filename"`
		Status    string `json:"status"`
		Additions string `json:"additions"`
		Deletions string `json:"deletions"`
		BlobURL   string `json:"blob_url"`
		RawURL    string `json:"raw_url"`
		Patch     struct {
			Diff        string `json:"diff"`
			NewPath     string `json:"new_path"`
			OldPath     string `json:"old_path"`
			AMode       string `json:"a_mode"`
			BMode       string `json:"b_mode"`
			NewFile     bool   `json:"new_file"`
			RenamedFile bool   `json:"renamed_file"`
			DeletedFile bool   `json:"deleted_file"`
			TooLarge    bool   `json:"too_large"`
		} `json:"patch"`
	}

	prCommit struct {
		URL         string `json:"url"`
		Sha         string `json:"sha"`
		HtmlURL     string `json:"html_url"`
		CommentsURL string `json:"comments_url"`
		Commit      struct {
			URL    string `json:"url"`
			Author struct {
				Name  string    `json:"name"`
				Date  time.Time `json:"date"`
				Email string    `json:"email"`
			} `json:"author"`
			Committer struct {
				Name  string    `json:"name"`
				Date  time.Time `json:"date"`
				Email string    `json:"email"`
			} `json:"committer"`
			Message      string `json:"message"`
			CommentCount int    `json:"comment_count"`
		} `json:"commit"`
		Author    author `json:"author"`
		Committer author `json:"committer"`
		Parents   struct {
			URL string `json:"url"`
			Sha string `json:"sha"`
		} `json:"parents"`
	}
)

type prInput struct {
	Title string `json:"title"`
	Body  string `json:"body"`
	Head  string `json:"head"`
	Base  string `json:"base"`
}
type prCommentInput struct {
	Body string `json:"body"`
}

func convertPullRequestList(from []*pr) []*scm.PullRequest {
	to := []*scm.PullRequest{}
	for _, v := range from {
		to = append(to, convertPullRequest(v))
	}
	return to
}
func convertPullRequest(from *pr) *scm.PullRequest {
	var labels []scm.Label
	for _, label := range from.Labels {
		labels = append(labels, scm.Label{
			Name:  label.Name,
			Color: label.Color,
		})
	}
	merged := from.State == "merged"
	closed := from.State == "closed"
	if merged {
		closed = true
	}
	return &scm.PullRequest{
		Number: from.Number,
		Title:  from.Title,
		Body:   from.Body,
		Sha:    from.Head.Sha,
		Ref:    fmt.Sprintf("refs/pull/%d/head", from.Number),
		Source: from.Head.Ref,
		Target: from.Base.Ref,
		Fork:   from.Head.Repo.FullName,
		Link:   from.HtmlURL,
		Diff:   from.DiffURL,
		Closed: closed,
		Merged: merged,
		Head: scm.Reference{
			Name: from.Head.Ref,
			Path: scm.ExpandRef(from.Head.Ref, "refs/heads"),
			Sha:  from.Head.Sha,
		},
		Base: scm.Reference{
			Name: from.Base.Ref,
			Path: scm.ExpandRef(from.Base.Ref, "refs/heads"),
			Sha:  from.Base.Sha,
		},
		Author: scm.User{
			Login:  from.User.Login,
			Avatar: from.User.AvatarURL,
		},
		Created: from.CreatedAt,
		Updated: from.UpdatedAt,
		Labels:  labels,
	}
}

func convertPullRequestComments(from []*prComment) []*scm.Comment {
	to := []*scm.Comment{}
	for _, v := range from {
		to = append(to, convertPullRequestComment(v))
	}
	return to
}
func convertPullRequestComment(from *prComment) *scm.Comment {
	return &scm.Comment{
		ID:   from.ID,
		Body: from.Body,
		Author: scm.User{
			Login:  from.User.Login,
			Name:   from.User.Name,
			Avatar: from.User.AvatarURL,
		},
		Created: from.CreatedAt,
		Updated: from.UpdatedAt,
	}
}

func convertPrChangeList(from []*prFile) []*scm.Change {
	to := []*scm.Change{}
	for _, v := range from {
		to = append(to, convertPrChange(v))
	}
	return to
}
func convertPrChange(from *prFile) *scm.Change {
	return &scm.Change{
		Path:    from.Filename,
		Added:   from.Status == "added",
		Deleted: from.Status == "deleted",
		Renamed: from.Status == "renamed",
		BlobID:  from.Sha,
	}
}

func convertPrCommitList(from []*prCommit) []*scm.Commit {
	to := []*scm.Commit{}
	for _, v := range from {
		to = append(to, convertPrCommit(v))
	}
	return to
}
func convertPrCommit(from *prCommit) *scm.Commit {
	return &scm.Commit{
		Message: from.Commit.Message,
		Sha:     from.Sha,
		Link:    from.HtmlURL,
		Author: scm.Signature{
			Name:   from.Commit.Author.Name,
			Email:  from.Commit.Author.Email,
			Date:   from.Commit.Author.Date,
			Login:  from.Author.Login,
			Avatar: from.Author.AvatarURL,
		},
		Committer: scm.Signature{
			Name:   from.Commit.Committer.Name,
			Email:  from.Commit.Committer.Email,
			Date:   from.Commit.Committer.Date,
			Login:  from.Committer.Login,
			Avatar: from.Committer.AvatarURL,
		},
	}
}
