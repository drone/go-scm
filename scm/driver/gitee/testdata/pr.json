{
  "id": 4730187,
  "url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6",
  "html_url": "https://gitee.com/kit101/drone-yml-test/pulls/6",
  "diff_url": "https://gitee.com/kit101/drone-yml-test/pulls/6.diff",
  "patch_url": "https://gitee.com/kit101/drone-yml-test/pulls/6.patch",
  "issue_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/issues",
  "commits_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/commits",
  "review_comments_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/comments/{/number}",
  "review_comment_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/comments",
  "comments_url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/6/comments",
  "number": 6,
  "state": "open",
  "title": "new-feature",
  "body": "Please pull these awesome changes",
  "assignees_number": 1,
  "testers_number": 1,
  "assignees": [],
  "testers": [],
  "milestone": null,
  "labels": [],
  "locked": false,
  "created_at": "2021-09-30T16:38:19+08:00",
  "updated_at": "2021-09-30T16:38:19+08:00",
  "closed_at": null,
  "merged_at": null,
  "mergeable": true,
  "can_merge_check": false,
  "head": {
    "label": "crud",
    "ref": "crud",
    "sha": "411db0d9e9072851383751c260dc1601840906e4",
    "user": {
      "id": 1535738,
      "login": "kit101",
      "name": "kit101",
      "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
      "url": "https://gitee.com/api/v5/users/kit101",
      "html_url": "https://gitee.com/kit101",
      "remark": "",
      "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
      "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
      "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
      "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
      "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
      "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
      "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
      "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
      "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
      "type": "User"
    },
    "repo": {
      "id": 14836026,
      "full_name": "kit101/drone-yml-test",
      "human_name": "kit101/drone-yml-test",
      "url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test",
      "namespace": {
        "id": 1576684,
        "type": "personal",
        "name": "kit101",
        "path": "kit101",
        "html_url": "https://gitee.com/kit101"
      },
      "path": "drone-yml-test",
      "name": "drone-yml-test",
      "owner": {
        "id": 1535738,
        "login": "kit101",
        "name": "kit101",
        "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
        "url": "https://gitee.com/api/v5/users/kit101",
        "html_url": "https://gitee.com/kit101",
        "remark": "",
        "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
        "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
        "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
        "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
        "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
        "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
        "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
        "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
        "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
        "type": "User"
      },
      "assigner": {
        "id": 1535738,
        "login": "kit101",
        "name": "kit101",
        "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
        "url": "https://gitee.com/api/v5/users/kit101",
        "html_url": "https://gitee.com/kit101",
        "remark": "",
        "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
        "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
        "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
        "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
        "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
        "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
        "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
        "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
        "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
        "type": "User"
      },
      "description": "",
      "private": false,
      "public": true,
      "internal": false,
      "fork": false,
      "html_url": "https://gitee.com/kit101/drone-yml-test.git",
      "ssh_url": "git@gitee.com:kit101/drone-yml-test.git"
    }
  },
  "base": {
    "label": "master",
    "ref": "master",
    "sha": "2eac1cac02c325058cf959725c45b0612d3e8177",
    "user": {
      "id": 1535738,
      "login": "kit101",
      "name": "kit101",
      "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
      "url": "https://gitee.com/api/v5/users/kit101",
      "html_url": "https://gitee.com/kit101",
      "remark": "",
      "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
      "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
      "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
      "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
      "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
      "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
      "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
      "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
      "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
      "type": "User"
    },
    "repo": {
      "id": 14836026,
      "full_name": "kit101/drone-yml-test",
      "human_name": "kit101/drone-yml-test",
      "url": "https://gitee.com/api/v5/repos/kit101/drone-yml-test",
      "namespace": {
        "id": 1576684,
        "type": "personal",
        "name": "kit101",
        "path": "kit101",
        "html_url": "https://gitee.com/kit101"
      },
      "path": "drone-yml-test",
      "name": "drone-yml-test",
      "owner": {
        "id": 1535738,
        "login": "kit101",
        "name": "kit101",
        "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
        "url": "https://gitee.com/api/v5/users/kit101",
        "html_url": "https://gitee.com/kit101",
        "remark": "",
        "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
        "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
        "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
        "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
        "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
        "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
        "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
        "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
        "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
        "type": "User"
      },
      "assigner": {
        "id": 1535738,
        "login": "kit101",
        "name": "kit101",
        "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
        "url": "https://gitee.com/api/v5/users/kit101",
        "html_url": "https://gitee.com/kit101",
        "remark": "",
        "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
        "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
        "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
        "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
        "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
        "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
        "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
        "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
        "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
        "type": "User"
      },
      "description": "",
      "private": false,
      "public": true,
      "internal": false,
      "fork": false,
      "html_url": "https://gitee.com/kit101/drone-yml-test.git",
      "ssh_url": "git@gitee.com:kit101/drone-yml-test.git"
    }
  },
  "_links": {
    "self": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6"
    },
    "html": {
      "href": "https://gitee.com/kit101/drone-yml-test/pulls/6"
    },
    "issue": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/issues"
    },
    "comments": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/comments"
    },
    "review_comments": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/comments"
    },
    "review_comment": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/comments/6"
    },
    "commits": {
      "href": "https://gitee.com/api/v5/repos/kit101/drone-yml-test/pulls/6/commits"
    }
  },
  "user": {
    "id": 1535738,
    "login": "kit101",
    "name": "kit101",
    "avatar_url": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png",
    "url": "https://gitee.com/api/v5/users/kit101",
    "html_url": "https://gitee.com/kit101",
    "remark": "",
    "followers_url": "https://gitee.com/api/v5/users/kit101/followers",
    "following_url": "https://gitee.com/api/v5/users/kit101/following_url{/other_user}",
    "gists_url": "https://gitee.com/api/v5/users/kit101/gists{/gist_id}",
    "starred_url": "https://gitee.com/api/v5/users/kit101/starred{/owner}{/repo}",
    "subscriptions_url": "https://gitee.com/api/v5/users/kit101/subscriptions",
    "organizations_url": "https://gitee.com/api/v5/users/kit101/orgs",
    "repos_url": "https://gitee.com/api/v5/users/kit101/repos",
    "events_url": "https://gitee.com/api/v5/users/kit101/events{/privacy}",
    "received_events_url": "https://gitee.com/api/v5/users/kit101/received_events",
    "type": "User"
  }
}