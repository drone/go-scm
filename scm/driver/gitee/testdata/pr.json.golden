{
  "Number": 6,
  "Title": "new-feature",
  "Body": "Please pull these awesome changes",
  "Sha": "411db0d9e9072851383751c260dc1601840906e4",
  "Ref": "refs/pull/6/head",
  "Source": "crud",
  "Target": "master",
  "Fork": "kit101/drone-yml-test",
  "Link": "https://gitee.com/kit101/drone-yml-test/pulls/6",
  "Diff": "https://gitee.com/kit101/drone-yml-test/pulls/6.diff",
  "Closed": false,
  "Merged": false,
  "Base": {
    "Sha": "2eac1cac02c325058cf959725c45b0612d3e8177",
    "Path": "refs/heads/master",
    "Name": "master"
  },
  "Head": {
    "Sha": "411db0d9e9072851383751c260dc1601840906e4",
    "Path": "refs/heads/crud",
    "Name": "crud"
  },
  "Author": {
    "Login": "kit101",
    "Avatar": "https://portrait.gitee.com/uploads/avatars/user/511/1535738_qkssk1711_1578953939.png"
  },
  "Created": "2021-09-30T16:38:19+08:00",
  "Updated": "2021-09-30T16:38:19+08:00"
}